kmir/src/kmir/kdist/mir-semantics/rt/data.md

```diff
@@ -1756,25 +1811,57 @@ The unary operation `unOpPtrMetadata`, when given a reference or pointer to a sl
 
   // could add a rule for cases without metadata
 ```
 
 
 
 #### Pointer Artithmetic
-Currently only supporting a trivial case where `binOpOffset` applies an offset of `0`, returning the same pointer.
+Addding an offset is currently restricted to unsigned values of an length, this may be too restrictive TODO Check.
+A pointer is offset by adding the magnitude of the `Integer` provided, as along as it is within the bounds of the pointer.
+It is valid to offset to the end of the pointer, however I believe it in not valid to read from there TODO: Check.
+A trivial case where `binOpOffset` applies an offset of `0` is added with higher priority as it is returning the same pointer.
 
 ```k
- rule #applyBinOp(
-         binOpOffset,
-         PtrLocal( STACK_DEPTH , PLACE , MUT, POINTEE_METADATA ),
-         Integer(0, _WIDTH, _SIGNED), // Trivial case when adding 0
-         _CHECKED)
-   =>
-         PtrLocal( STACK_DEPTH , PLACE , MUT, POINTEE_METADATA )
+  // Trivial case when adding 0 - valid for any pointer
+  rule #applyBinOp(
+          binOpOffset,
+          PtrLocal( STACK_DEPTH , PLACE , MUT, POINTEE_METADATA ),
+          Integer(0, _WIDTH, _SIGNED), // Trivial case when adding 0
+          _CHECKED)
+    =>
+          PtrLocal( STACK_DEPTH , PLACE , MUT, POINTEE_METADATA )
+   [preserves-definedness, priority(40)]
+
+  // Check offset bounds against origin pointer with dynamicSize metadata
+  rule #applyBinOp(
+          binOpOffset,
+          PtrLocal( STACK_DEPTH , PLACE , MUT, ptrEmulation(CURRENT_META, CURRENT_OFFSET, 
+                    PtrLocal(ORIG_DEPTH, ORIG_PLACE, ORIG_MUT, ptrEmulation(dynamicSize(ORIG_LENGTH), ORIG_OFFSET, ORIG_ORIGIN))) ),
+          Integer(OFFSET_VAL, _WIDTH, false), // unsigned offset
+          _CHECKED)
+    =>
+          PtrLocal( STACK_DEPTH , PLACE , MUT, ptrEmulation(CURRENT_META, CURRENT_OFFSET +Int OFFSET_VAL, 
+                    PtrLocal(ORIG_DEPTH, ORIG_PLACE, ORIG_MUT, ptrEmulation(dynamicSize(ORIG_LENGTH), ORIG_OFFSET, ORIG_ORIGIN))) )
+    requires OFFSET_VAL >=Int 0
+     andBool CURRENT_OFFSET +Int OFFSET_VAL <=Int ORIG_LENGTH
+   [preserves-definedness]
+
+  // Check offset bounds against origin pointer with staticSize metadata
+  rule #applyBinOp(
+          binOpOffset,
+          PtrLocal( STACK_DEPTH , PLACE , MUT, ptrEmulation(CURRENT_META, CURRENT_OFFSET, 
+                    PtrLocal(ORIG_DEPTH, ORIG_PLACE, ORIG_MUT, ptrEmulation(staticSize(ORIG_LENGTH), ORIG_OFFSET, ORIG_ORIGIN))) ),
+          Integer(OFFSET_VAL, _WIDTH, false), // unsigned offset
+          _CHECKED)
+    =>
+          PtrLocal( STACK_DEPTH , PLACE , MUT, ptrEmulation(CURRENT_META, CURRENT_OFFSET +Int OFFSET_VAL, 
+                    PtrLocal(ORIG_DEPTH, ORIG_PLACE, ORIG_MUT, ptrEmulation(staticSize(ORIG_LENGTH), ORIG_OFFSET, ORIG_ORIGIN))) )
+    requires OFFSET_VAL >=Int 0
+     andBool CURRENT_OFFSET +Int OFFSET_VAL <=Int ORIG_LENGTH
    [preserves-definedness]
```

Comment:
This looks like we don't have to store the entire `PtrLocal` as the origin, we only need to know the `ORIG_LENGTH`. That length is stored as `Metadata`.
What happens when offsetting a pointer to something not a slice or array? If we use the metadata, then we get that original metadata is `noMetadata` . In this version here we probably never find a non-pointer in the origin field?.